test/integration/vshard_custom_test.lua

```diff
@@ -0,0 +1,1660 @@
+                local router = router_service.get(opts.vshard_router)
+                opts.vshard_router = router
+            end
+
+            local result = {}
+            for _, v in crud.pairs(space_name, arg1, opts) do
+                table.insert(result, v)
+            end
+
+            return result
+        end
+
+        rawset(_G, 'call_pairs_wrapper', call_pairs_wrapper)
+    ]])
+
+    g.call_router_opts2 = function(g, ...)
+        return g.router:call('call_wrapper_opts2', {g.params.option, ...})
+    end
+
+    g.call_router_opts3 = function(g, ...)
+        return g.router:call('call_wrapper_opts3', {g.params.option, ...})
+    end
+
+    g.call_router_opts4 = function(g, ...)
+        return g.router:call('call_wrapper_opts4', {g.params.option, ...})
+    end
+
+    g.call_router_pairs = function(g, ...)
+        return g.router:call('call_pairs_wrapper', {g.params.option, ...})
+    end
+end)
+
+pgroup.before_each(function(g)
+    helpers.truncate_space_on_cluster(g.cluster, 'customers')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_ddl')
+    helpers.truncate_space_on_cluster(g.cluster, 'locations')
+    helpers.truncate_space_on_cluster(g.cluster, 'locations_ddl')
+
+    g.router:call('prepare_data', {'customers', 'customers', {1, box.NULL, 'Akiyama Shun', 32}})
+    g.router:call('prepare_data', {'customers', 'customers', {2, box.NULL, 'Kazuma Kiryu', 41}})
+    g.router:call('prepare_data', {'customers_ddl', 'customers', {1, box.NULL, 'Akiyama Shun', 32}})
+    g.router:call('prepare_data', {'customers_ddl', 'customers', {2, box.NULL, 'Kazuma Kiryu', 41}})
+
+    g.router:call('prepare_data', {'locations', 'locations', {'Sky Finance', box.NULL, 'Credit company', 2}})
+    g.router:call('prepare_data', {'locations', 'locations', {'Sunflower', box.NULL, 'Orphanage', 1}})
+    g.router:call('prepare_data', {'locations_ddl', 'locations', {'Sky Finance', box.NULL, 'Credit company', 2}})
+    g.router:call('prepare_data', {'locations_ddl', 'locations', {'Sunflower', box.NULL, 'Orphanage', 1}})
+end)
+
+pgroup.test_call_min = function(g)
+    local result, err = g:call_router_opts3(
+        'min', 'customers', 'age', {vshard_router = 'customers'})
+
+    t.assert_equals(err, nil)
+    t.assert_items_equals(result.rows, {{1, 12477, 'Akiyama Shun', 32}})
+end
+
+pgroup.test_call_min_wrong_router = function(g)
+    local result, err = g:call_router_opts3(
+        'min', 'customers', 'age', {vshard_router = 'locations'})
+
+    t.assert_equals(result, nil)
+    t.assert_str_contains(err.err, "Space \"customers\" doesn't exist")
+end
+
+pgroup.test_call_min_no_router = function(g)
+    local result, err = g:call_router_opts3(
+        'min', 'customers', 'age')
+
+    t.assert_equals(result, nil)
+    t.assert_str_contains(err.err,
+                          "Default vshard group is not found and custom is not specified with opts.vshard_router")
+end
```

Comment:
I think we can use a table of cases here:

```Lua
local call_cases = {
   min = {
       call = "min",
       opts = {'customers', 'age'}
       expected = {{1, 12477, 'Akiyama Shun', 32}}
   },
   select = {
       ...
   },
   ...
}

for name, tc in ipairs(call_cases) do
   pgroup["test_" .. name] = function(g)
       ...
       t.assert_equals(err, nil)
       t.assert_items_equals(result.rows, tc.expected)
   end
   pgroup["test_" .. name .. "_no_router"] = function(g)
       ...
       t.assert_equals(result, nil)
       t.assert_str_contains(err.err,
                          "Default vshard group is not found and custom is not specified with opts.vshard_router")
   end
   pgroup["test_" .. name .. "_wrong_option"] = function(g)
       ...
       t.assert_equals(result, nil)
       t.assert_str_contains(err.err,
                          "Invalid opts.vshard_router table value, a vshard router instance has been expected")
   end
end
```

There will be less the repeating code.

Reply:
I have thought about building test cases here too (especially for `wrong_router`/`no_router`/`wrong_option`).

The reason I've used plain copypaste is because test cases are really small and introducing more layers won't make test more compact but will make it less readable. But this approach is still prone to human error (if I have forhot to change something in some case).